tests/trainer/logging_/test_eval_loop_logging.py

```diff
@@ -672,3 +673,29 @@ def val_dataloader(self):
         enable_model_summary=False,
     )
     trainer.fit(model)
+
+
+@pytest.mark.parametrize("num_dataloaders", [1, 2, 11])
+def test_log_metrics_only_include_metrics_from_concerned_dataloader(num_dataloaders, tmpdir):
+    class TestModel(BoringModel):
+        def test_step(self, batch, batch_idx, dataloader_idx=0):
+            output = self.layer(batch)
+            loss = self.loss(batch, output)
+            self.log("fake_test_acc", loss)
+            return {"y": loss}
+
+        def test_epoch_end(self, *_) -> None:
+            pass
+
+    test = RandomDataset(32, 2)
+    test_dataloaders = [DataLoader(test, batch_size=1)] * num_dataloaders
+
+    model = TestModel()
+    trainer = Trainer(default_root_dir=tmpdir)
+
+    output = trainer.test(model, dataloaders=test_dataloaders)
+    assert sum(len(x) for x in output) == num_dataloaders
+    if num_dataloaders == 1:
+        assert "dataloader_idx" not in output[0]
+    else:
+        assert all(f"dataloader_idx_{idx}" == list(x.keys())[0].split("/")[1] for idx, x in enumerate(output))
```

Comment:
This assert will fail for precisely the bug that led to this PR:

```python
wrong_output = [{'blah/dataloader_idx_0': 'dummy'}, {'blah/dataloader_idx_1': 'dummy', 'blah/dataloader_idx_10': 'dummy'}]

# raises no error
assert all(f"dataloader_idx_{idx}" == list(x.keys())[0].split("/")[1] for idx, x in enumerate(wrong_output))
````

I'll try to come up with a better test that is more specific tomorrow :)

Reply:
I've tried to re-write the test to make it more specific. I am not sure how much better it is than my initial very simple test.